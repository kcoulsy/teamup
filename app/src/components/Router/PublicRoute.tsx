import React from 'react';
import { Route, Redirect, RouteProps } from 'react-router-dom';
import * as H from 'history';
import useUser from '../../hooks/useUser';

export const PublicRoute: React.FC<RouteProps> = ({ children, ...rest }) => {
  const { isLoggedIn } = useUser();
  //@ts-ignore
  return <Route {...rest} render={publicRouteRender(isLoggedIn, children)} />;
};

export interface PublicRouteRenderProps {
  location: H.Location<any>;
}

export const publicRouteRender = (
  isLoggedIn: boolean,
  children: React.ReactNode
) => {
  return (props: PublicRouteRenderProps): React.ReactNode => {
    if (isLoggedIn) {
      return (
        <Redirect to={{ pathname: '/', state: { from: props.location } }} />
      );
    }

    return children;
  };
};

export default PublicRoute;
