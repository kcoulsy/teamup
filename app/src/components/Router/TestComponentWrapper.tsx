import React from 'react';

const TestComponentWrapper: React.FunctionComponent = ({ children }) => {
  return <div>{children}</div>;
};

export default TestComponentWrapper;
