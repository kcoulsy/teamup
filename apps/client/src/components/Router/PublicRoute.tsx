import React from 'react';
import { Route, Redirect, RouteProps } from 'react-router-dom';
import * as H from 'history';
import useUser from '../../hooks/useUser';

export const publicRouteRender = (isLoggedIn: boolean, children: React.ReactNode) =>
  function redirectComponent({ location }: PublicRouteRenderProps): React.ReactNode {
    if (isLoggedIn) {
      return <Redirect to={{ pathname: '/', state: { from: location } }} />;
    }

    return children;
  };

function PublicRoute({ children, ...rest }: RouteProps) {
  const { isLoggedIn } = useUser();
  // @ts-ignore
  return <Route {...rest} render={publicRouteRender(isLoggedIn, children)} />;
}

export interface PublicRouteRenderProps {
  location: H.Location<any>;
}

export default PublicRoute;
