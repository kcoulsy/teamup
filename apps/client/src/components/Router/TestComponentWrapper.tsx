import React, { PropsWithChildren } from 'react';

function TestComponentWrapper({ children }: PropsWithChildren<{}>) {
  return <div>{children}</div>;
}

export default TestComponentWrapper;
