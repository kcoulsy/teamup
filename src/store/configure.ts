import { configureStore } from '@reduxjs/toolkit';

// Import Reducers
import authReducer from '../reducers/auth';
import teamReducer from '../reducers/team';

const store = configureStore({
    reducer: {
        // reducers here
        auth: authReducer,
        team: teamReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
