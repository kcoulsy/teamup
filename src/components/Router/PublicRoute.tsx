import React from 'react';
import { Route, Redirect, RouteProps } from 'react-router-dom';
import { connect } from 'react-redux';
import isLoggedIn from '../../helpers/isLoggedIn';
import { RootState } from '../../store/configure';
import * as H from 'history';

interface PublicRouteProps extends RouteProps {
  isLoggedIn: boolean;
}

export const PublicRoute: React.FunctionComponent<PublicRouteProps> = ({
  isLoggedIn,
  children,
  ...rest
}) => {
  //@ts-ignore
  return <Route {...rest} render={publicRouteRender(isLoggedIn, children)} />;
};

export interface PublicRouteRenderProps {
  location: H.Location<any>;
}

export const publicRouteRender = (
  isLoggedIn: boolean,
  children: React.ReactNode
) => {
  return (props: PublicRouteRenderProps): React.ReactNode => {
    if (isLoggedIn) {
      return (
        <Redirect to={{ pathname: '/', state: { from: props.location } }} />
      );
    }

    return children;
  };
};

export const mapStateToProps = (state: RootState) => {
  return {
    isLoggedIn: isLoggedIn(state.auth.token),
  };
};

export default connect(mapStateToProps)(PublicRoute);
