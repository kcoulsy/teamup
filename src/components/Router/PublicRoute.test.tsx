import React from 'react';
import { shallow } from 'enzyme';

import { PublicRoute, publicRouteRender, mapStateToProps } from './PublicRoute';
import { shallowToJson } from 'enzyme-to-json';
import isLoggedIn from '../../helpers/isLoggedIn';
import * as H from 'history';
import { RootState } from '../../store/configure';
import TestComponentWrapper from './TestComponentWrapper';

describe('render PublicRoute component', () => {
  it('should render PublicRoute component', () => {
    const wrapper = shallow(<PublicRoute isLoggedIn={false} />);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should handle publicRouteRender when logged in', () => {
    const child = <p>JSX Child</p>;
    const isLoggedIn = true;
    const location: H.Location<any> = {
      pathname: '',
      search: '',
      state: '',
      hash: '',
      key: '',
    };

    const wrapper = shallow(
      <TestComponentWrapper>
        {publicRouteRender(isLoggedIn, child)({ location })}
      </TestComponentWrapper>
    );

    expect(wrapper.contains(child)).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle publicRouteRender when not logged in', () => {
    const child = <p>JSX Child</p>;
    const isLoggedIn = false;
    const location: H.Location<any> = {
      pathname: '',
      search: '',
      state: '',
      hash: '',
      key: '',
    };
    const wrapper = shallow(
      <TestComponentWrapper>
        {publicRouteRender(isLoggedIn, child)({ location })}
      </TestComponentWrapper>
    );

    expect(wrapper.contains(child)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should return correct props from store', () => {
    const mockState: RootState = {
      auth: {
        attemptingLogin: false,
        token: 'testtoken',
        loginAttemptFailed: false,
        appInitialising: false,
        appInitialised: false,
        attemptingRegister: false,
      },
    };

    expect(mapStateToProps(mockState)).toEqual({
      isLoggedIn: isLoggedIn(mockState.auth.token),
    });
  });
});
