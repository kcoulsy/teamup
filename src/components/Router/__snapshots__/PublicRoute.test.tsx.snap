// Vitest Snapshot v1

exports[`render PublicRoute component > should handle publicRouteRender when logged in 1`] = `ShallowWrapper {}`;

exports[`render PublicRoute component > should handle publicRouteRender when not logged in 1`] = `ShallowWrapper {}`;

exports[`render PublicRoute component > should render PublicRoute component 1`] = `
<Route
  render={[Function]}
/>
`;

exports[`render PublicRoute component should handle publicRouteRender when logged in 1`] = `
<div>
  <Redirect
    to={
      Object {
        "pathname": "/",
        "state": Object {
          "from": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": "",
          },
        },
      }
    }
  />
</div>
`;

exports[`render PublicRoute component should handle publicRouteRender when not logged in 1`] = `
<div>
  <p>
    JSX Child
  </p>
</div>
`;

exports[`render PublicRoute component should render PublicRoute component 1`] = `
<Route
  render={[Function]}
/>
`;
